keyring/keyring.go

```diff
@@ -0,0 +1,793 @@
+		// For p2tr BIP0086 key spend only.
+		case taprootKeySpendBIP0086SignMethod:
+			rootHash := make([]byte, 0)
+			err = k.signSegWitV1KeySpend(
+				in, tx, sigHashes, idx, rootHash,
+			)
+
+		// For p2tr with script commitment key spend path.
+		case taprootKeySpendSignMethod:
+			rootHash := in.TaprootMerkleRoot
+			err = k.signSegWitV1KeySpend(
+				in, tx, sigHashes, idx, rootHash,
+			)
+
+		// For p2tr script spend path.
+		case taprootScriptSpendSignMethod:
+			leafScript := in.TaprootLeafScript[0]
+			leaf := txscript.TapLeaf{
+				LeafVersion: leafScript.LeafVersion,
+				Script:      leafScript.Script,
+			}
+			err = k.signSegWitV1ScriptSpend(
+				in, tx, sigHashes, idx, leaf,
+			)
+
+		default:
+			err = fmt.Errorf("unsupported signing method for "+
+				"PSBT signing: %v", signMethod)
+		}
+		if err != nil {
+			return nil, err
+		}
+
+		signedInputs = append(signedInputs, uint32(idx))
+
+		/* newTx := packet.UnsignedTx.Copy()
+		newTx.TxIn[0].SignatureScript = in.RedeemScript
+		newTx.TxIn[0].Witness = wire.TxWitness{
+			in.PartialSigs[0].Signature,
+			in.PartialSigs[0].PubKey,
+		}
+
+		log.Infof("Executing engine on tx input %+v (from %+v)",
+			newTx.TxIn[0], in)
+
+		engine, err := txscript.NewEngine(
+			in.WitnessUtxo.PkScript,
+			newTx,
+			idx,
+			txscript.StandardVerifyFlags,
+			txscript.NewSigCache(10),
+			sigHashes,
+			in.WitnessUtxo.Value,
+			prevOutputFetcher,
+		)
+		if err != nil {
+			log.Errorf("Error creating engine: %v", err)
+			continue
+		}
+
+		err = engine.Execute()
+		if err != nil {
+			log.Errorf("Error executing engine: %v", err)
+			continue
+		}
+
+		log.Infof("Succeeded executing engine") */
+	}
+
+	return signedInputs, nil
+}
+
+// signSegWitV0 attempts to generate a signature for a SegWit version 0 input
+// and stores it in the PartialSigs (and FinalScriptSig for np2wkh addresses)
+// field.
+func (k *KeyRing) signSegWitV0(in *psbt.PInput, tx *wire.MsgTx,
+	sigHashes *txscript.TxSigHashes, idx int) error {
+
+	if len(in.Bip32Derivation) == 0 {
+		return nil
+	}
+
+	// Extract the correct witness and/or legacy scripts now, depending on
+	// the type of input we sign. The txscript package has the peculiar
+	// requirement that the PkScript of a P2PKH must be given as the witness
+	// script in order for it to arrive at the correct sighash. That's why
+	// we call it subScript here instead of witness script.
+	subScript := prepareScriptsV0(in)
+
+	// We have everything we need to calculate the digest for signing.
+	digest, err := txscript.CalcWitnessSigHash(subScript, sigHashes,
+		in.SighashType, tx, idx, in.WitnessUtxo.Value)
+	if err != nil {
+		return fmt.Errorf("error getting sighash for input %d: %v",
+			idx, err)
+	}
+
+	log.Tracef("Got input %+v for signing with unknowns %+v", in,
+		in.Unknowns)
+
+	reqData := map[string]interface{}{
+		"node":   k.node,
+		"path":   sliceUint32ToInt(in.Bip32Derivation[0].Bip32Path),
+		"method": "ecdsa",
+		"digest": hex.EncodeToString(digest),
```

Comment:
I suppose a main decision to make here is whether to pass the full psbt to vault. Probably need to chat about that a bit more, also in conjunction with policies.